### ISSUE TITLE
Simtime to double conversion prevents correct comparison of IEEE 802.11 rates

### ISSUE THREAD

**Issue by lbobr** [inet-framework/inet#283]
`Ieee80211OFDMModeBase::computeGrossBitrate() `calculates actual gross bitrate via 
`codedBitsPerOFDMSymbol / getSymbolInterval()` which yields non-accurate numbers. As a result comparison of rates in `Ieee80211ModeSet::findMode()` fails.

The following works for me:

```
--- src/inet/physicallayer/ieee80211/mode/Ieee80211OFDMMode.cc
+++ src/inet/physicallayer/ieee80211/mode/Ieee80211OFDMMode.cc
@@ -61,7 +61,7 @@ Ieee80211OFDMDataMode::Ieee80211OFDMDataMode(const Ieee80211OFDMCode *code, cons
 bps Ieee80211OFDMModeBase::computeGrossBitrate(const Ieee80211OFDMModulation *modulation) const
 {
     int codedBitsPerOFDMSymbol = modulation->getSubcarrierModulation()->getCodeWordSize() * getNumberOfDataSubcarriers();
-    return bps(codedBitsPerOFDMSymbol / getSymbolInterval());
+    return bps((codedBitsPerOFDMSymbol / ((double)getSymbolInterval().raw() / getSymbolInterval().getScale())));
 }
```

**Comment by levy** [comment]
Could you give me an example?

**Comment by lbobr** [comment]
In my case, i am specifying a certain opMode (e.g. 'p') and bitrate (e.g. '12Mbps') for a Ieee80211Radio via omnetpp.ini. Subsequently, the model tries to find the correct mode during initialization. My simtime resulution is thereby set to 1ns. While debugging, I saw that not the exact bitrate is returned but rather something lie, for example, "11.99999995". If i use 1 ps as a resulution, it works as expected and "12" is returned.

**Comment by levy** [comment]
Thank you, I have reproduced this problem. Unfortunately, I'm not sure if your patch is right as it is now:
 - after applying the patch its unclear why the code is written the way it is, so it will be tempting to "fix it" and reintroduce the bug later
 - the bug is actually in SimTime::operator/() it simply converts the simulation time to a double and divide the integer with the double. In this case, and probably in many other cases this could be done without precision loss. E.g. we could multiply with the inverse scale first and divide with raw value, check if there are overflow, etc.

Maybe the best would be to introduce a new division operation into SimTime and use that in the Ieee80211ModeSet, at least it would give a hint for the future maintainers not to "fix it".

**Comment by levy** [comment]
I reported this bug in the OMNeT++ bug tracker as #0001038.

I close this issue for two reasons:
 - the bug is already fixed in the 802.11 implementation. The fix allows the bitrate to match only 0.1 Mhz precision. This is already needed for bitrates which are infinite numbers, e.g. 86,666... or the like
 - the bug is alread reported in the OMNeT++ bug tracker
